### ISSUE TITLE
ghc-lib-parser-8.10.7.20210828 build error with ghc-9.0.1

### ISSUE THREAD

**Issue by jneira** [digital-asset/ghc-lib#332]
Hi, after updating the hackage index and invalidate the cache of haskell-language-server ci builds we have detected a build error:

```
[ 65 of 228] Compiling Util             ( compiler/utils/Util.hs, dist/build/Util.o, dist/build/Util.dyn_o )

compiler/utils/Util.hs:43:28: error:
Error:     Ambiguous occurrence ‘singleton’
    It could refer to
       either ‘Data.List.singleton’,
              imported from ‘Data.List’ at compiler/utils/Util.hs:141:1-38
              (and originally defined in ‘base-4.15.0.0:Data.OldList’)
           or ‘Util.singleton’, defined at compiler/utils/Util.hs:558:1
   |
43 |         isSingleton, only, singleton,
   |                            ^^^^^^^^^
```

https://github.com/jneira/haskell-language-server/runs/4187119335?check_suite_focus=true

Not sure what is the reason but it seems a pure internal compile error, independent of other dependencies.
I think we were able to build some ghc-lib-parser-8.10.* with ghc-9.0.1 at some point so i suppose it is possible.
I am afraid that it is blocking [the hackage release of haskell-language-server for ghc-9.0.1](https://github.com/haskell/haskell-language-server/issues/2281)

@shayne-fletcher Thanks in advance!

**Comment by jneira** [comment]
Building the package in isolation with ghc-9.0.1 also fails, as expected. The build is successful with ghc < 9
0.
However building the last 8.10.6 version with ghc-9.0.1 is also succesful.

**Comment by jneira** [comment]
> However building 8.10.6.20210814 version with ghc-9.0.1 is also succesful.

I've forced hls to get that version and also fails with the same error so my test may be wrong: https://github.com/jneira/haskell-language-server/runs/4208840518?check_suite_focus=true

I assume ghc-lib-parse >= 8.10.5.20210606 is buildable with ghc-9.0.1 as their base bounds allow it.

**Comment by jneira** [comment]
And it fails for 8.10.5.20210606 as well: https://github.com/jneira/haskell-language-server/runs/4209784703?check_suite_focus=true

So the unique hypothesis i have is we were not building hls for any ghc-lib-parser-8.10 and the recent inclusion of plugins using ghc-lib-parser have triggered the use of that version

**Comment by shayne-fletcher** [comment]
> I assume ghc-lib-parse >= 8.10.5.20210606 is buildable with ghc-9.0.1 as their base bounds allow it.

sorry @jneira , it's not. generally, we only support building "forward" (or rather, this is a ghc restriction). i'm sorry the bounds didn't prevent this. we're trying to tighten them up to exclude these cases. if you can see what we might have done to prevent a build plan here, i'd very much welcome a patch

**Comment by jneira** [comment]
thanks, i was afraid that it is the case, i would argue docs also are little bit ambiguous about: 

https://github.com/digital-asset/ghc-lib#why-doesnt-ghc-lib-version-x-build-with-ghc-compiler-version-y

does not mention the only "forward" compat explicitly and the table suggests that you can build a ghc-lib version with any superior ghc

**Comment by shayne-fletcher** [comment]
> thanks, i was afraid that it is the case, i would argue docs also are little bit ambiguous about:
> 
> https://github.com/digital-asset/ghc-lib#why-doesnt-ghc-lib-version-x-build-with-ghc-compiler-version-y
> 
> does not mention the only "forward" compat explicitly and the table suggests that you can build a ghc-lib version with any superior ghc

i agree. i'll try to improve it. i'm sorry for your inconvenience.

**Comment by jneira** [comment]
dont worry at all, in my end i should not have assumed ghc-lib would magically add forward compatibility. Seems almost obvious in retrospective :facepalm:

**Comment by jneira** [comment]
i think this can be closed, thanks for #366!
